### ISSUE TITLE
Document and support modular Flutter engines

### ISSUE THREAD

**Issue by xster** [flutter/flutter#40220]
For customers who critically care about the artifact size and don't mind spending time fine tuning a custom engine build to trim unneeded features, we should support a semi-convenient workflow. 

@dnfield also floated some ideas previously. 

Letting the user turn flags on and off via gn files manually and building local engines is likely ok as long as they're documented and runtime errors are properly surfaced. 

**For readers:** your feedback is also extremely valued (in helping us determine how to slice and dice the engine for optional modularity). To see the current engine size breakdown, please see https://ci.chromium.org/p/flutter/g/engine/console, click on the latest Linux/AOT build. Under `gsutil upload treemap for out/android_release_arm64/libflutter.so`, click Open Treemap. 

Example from 2019-09-11: https://storage.googleapis.com/flutter_infra/flutter/e2cc04c3fc74473cc3f2fbf45ae02e72b0f3594e/android-arm64-release/sizes/index.html

Each box in the graph is also double-clickable to further drill down into the contents. (cc @goderbauer to proof-read instructions)

**Comment by dnfield** [comment]
Relevant parts of the linked doc:

# Modularizing the Flutter Engine
I spent about a day fiddling with modularizing the Flutter engine, with a few goals in mind:

1. Reduce binary size for customers who don’t need certain features (e.g. Hello World doesn’t need image decoding libraries or boring SSL, which together make up ~half a meg uncompressed)
2. Enable customers who want more features exposed from Skia to use them (e.g. certain effects we don’t expose, or Skottie) at an increased binary cost - in other words, potentially increase the API surface we expose from Skia.

My goal was to see if building Skia and/or parts of Skia as dynamic libraries would allow us to do this.

## Challenges

- Need to develop a way to let consumers pick which modules they need.
- Need to determine how to fail at runtime if someone uses API that they don’t have a module for.
- Need to determine how to test the various combinations that this would involve.
- It may not be a good idea to expose all of Skia’s API surface. This won’t work on some platforms at all, and even on platforms where it works it may cause strange difficult to diagnose issues.
- Exposing more of Skia increases the binary size.
- We lose a lot of benefits of LTO, particularly for Skia and libcxx (I didn’t test Dart or Boring SSL but I assume the same would be true there).
- Each dynamic library ends up with additional overhead in file size.
- Even statically linking Skia without JPEG or WEBP support only takes off a very small amount of binary size - and since freetype needs libpng, we can’t really get rid of that.

If we could somehow get to a point where we have ABI stability for Dart, Engine, and Skia, we could explore other options for this that may have additional tradeoffs and only work on some platforms.

**Comment by dnfield** [comment]
I think a major challenge here for customers is that building the engine is non-trival and time consuming (particularly without GOMA or a really beefy machine, and particularly building release mode with LTO).

The testing/support issue is a big one in my mind - if you build your own engine and something crashes, what kind of support can you expect to get from the Flutter team?

**Comment by xster** [comment]
I think it's ok for this to be an iterative process where our bigger/more adventurous users can first try a slightly more hands-on approach that unblocks also the highest MAU apps from shipping. More convenience and lower barrier to entry for smaller scale apps is a linear progress after that.

**Comment by dnfield** [comment]
Another trick to it is we have no good way to conditionally include or exclude things from dart:ui based on build flags.

**Comment by jonahwilliams** [comment]
Please don't consider changing the API of dart:ui based on build flags.

**Comment by xster** [comment]
Thinking in very incremental steps: as an MVP, letting it crash at runtime and supplying some sort of linter which check application side API usage against a map seems ok to start with (and not alter any dart:ui APIs).

**Comment by dnfield** [comment]
I could vaguely imagine something like building your own dart:ffi module on top of a custom built engine being useful to someone.  Maybe with some kind of optional dart:ffi way of exposing more raw stuff in the engine and available as an optional module. But testing it would be pretty painful, without tests it'd be super fragile and likely just not work in a lot of important ways, and it would never even possibly work on at least two of the platforms we want to support - and even on platforms we do, there's a ton of stuff you'd be trying to dig into that are private implementation details right now that aren't guarnateed to work the same way (or even be present) from one commit to another on the platforms where it does work.

It'd be helpful to have a concrete use case of what we're trying to solve here though - my experiments with trying to modularize the engine all led to a path that said "this is a bad idea". @xster do you have any specific requests/use cases in mind for this?

**Comment by zzjjob** [comment]
We removed the skia, but the final size is not much reduced. 

In fact, the dart product is also quite big.

Engine build & test is tormenting,  In fact, everybody may use a different CI/CD.

**Comment by xster** [comment]
> In fact, the dart product is also quite big.

That's a slightly separate issue (since the AOT outputs is a separate so/dylib). Can you share some numbers so we can prioritize focusing on the engine size vs the Dart output size?

**Comment by dnfield** [comment]
We very heavily tuned and optimized our build flags.  When you removed skia, did you use the same flags?

Also, removing skia reduces the amount we can trim when you later link, since linking with a dynamic skia lib will mean we can't lto as well anymore.

**Comment by zzjjob** [comment]
Yes,  skia use the same build flags。But flutter engine skia build flag, 

Base flutter v1.7.8 armv7a release, ELF file zise:
libflutter.so : 5.8M -> 4.0M
libskia.so：2.8M

**Comment by GHHX** [comment]
Can you simplify the function? Keep only the basic functions。The rest of the function is made into expandable modules。
For example:
Skia engine without JPEG or GIF or WEBP，without Module Animator of Skia etc。
Flutter engine without Flutter for web。
Streamlining Rootisolate，Some functions that interact with the platform are made into modules。

**Comment by dnfield** [comment]
Removing the image codecs doesn't save a ton of space.  Maybe 50-100kb.

We don't include an animation module.

Flutter web is completely separate already 

What parts of root isolate do you have in mind?

**Comment by wangying3426** [comment]
@xster  Modularizing the Flutter Engine is really helpful for us, is there any progress or next plan? Thank you.

**Comment by dnfield** [comment]
@wangying3426 - it would be very helpful to have some more information to make decisions about how to proceed here.

What are your primary goals?  For example, are you most concerned with reducing disk footprint of the engine, reducing latency of startup at runtime, including features not currently bundled into the engine, or something else?

If you want to reduce the disk footprint, what features are you willing to sacrifice?  How much disk savings would you need to see to make these features be dispensible?

For example, removing an unused image decoder saves roughly 40-60kb uncompressed.  Compressed that goes down by roughly half. I would expect that most customers would rather keep the 20-30kb to support an image format they don't currently use but may use in the future.

But, of course, you may have other features or use cases in mind that I'm not thinking of :)

**Comment by wangying3426** [comment]
> @wangying3426 - it would be very helpful to have some more information to make decisions about how to proceed here.
> 
> What are your primary goals? For example, are you most concerned with reducing disk footprint of the engine, reducing latency of startup at runtime, including features not currently bundled into the engine, or something else?
> 
> If you want to reduce the disk footprint, what features are you willing to sacrifice? How much disk savings would you need to see to make these features be dispensible?
> 
> For example, removing an unused image decoder saves roughly 40-60kb uncompressed. Compressed that goes down by roughly half. I would expect that most customers would rather keep the 20-30kb to support an image format they don't currently use but may use in the future.
> 
> But, of course, you may have other features or use cases in mind that I'm not thinking of :)

First, we want to drop the Http/WebSocket library (include dart:http、dart:io、boringSSL) , because we must use our own network library by flutter plugin. Second, we will use our own libskia library, so the Flutter's is unnecessary, and will drop libwebp too.

**Comment by wangying3426** [comment]
> > @wangying3426 - it would be very helpful to have some more information to make decisions about how to proceed here.
> > What are your primary goals? For example, are you most concerned with reducing disk footprint of the engine, reducing latency of startup at runtime, including features not currently bundled into the engine, or something else?
> > If you want to reduce the disk footprint, what features are you willing to sacrifice? How much disk savings would you need to see to make these features be dispensible?
> > For example, removing an unused image decoder saves roughly 40-60kb uncompressed. Compressed that goes down by roughly half. I would expect that most customers would rather keep the 20-30kb to support an image format they don't currently use but may use in the future.
> > But, of course, you may have other features or use cases in mind that I'm not thinking of :)
> 
> @dnfield Firstly, we want to drop the Http/WebSocket library, include dart:http、dart:io、boringSSL, because we must use our own network library through flutter plugins for some apps, and it's selectively. Secondly, we will use our own libskia, so the Flutter's is unnecessary, and want to use our own libwebp too. Lastly, want to replace icu library if possible.
> 
> Could you have any suggestions about it ？ :) Thank you~

It seems that icu library can not  be replaced or tailored. 😌

**Comment by dnfield** [comment]
Sorry - I've been on leave and a bit slow to get back to things.

Using your own libskia may or may not introduce issues - Flutter uses at least one instance of non-public API from Skia (for base32 decoding ?!), and may use others.  Even for the public API, Flutter tends to build against the very latest revision of Skia, which sometimes has breaking changes to API that get fixed in Flutter (for example, when a Skia method changes its signature).

It's definitely possible to build Skia separately as a dynamic library, but it would likely be important to use the _exact same revision as what Flutter expects_ in the rest of your application, which may or may not help you here.

It would be interesting to have a way to let dart:io delegate to some other implementation rather than the Dart VM's.  I made comments more about this in #32161. I think to make it really work, you'd want to covery the entire API surface of dart:io w.r.t. Sockets and HTTP calls though - so that any other plugins or framework bits you use that rely on it will still work.

ICU has already been heavily optimized for Flutter - we've cut out a lot that we didn't need (/cc @goderbauer).  There are additional plans to see if we can compress some of the larger chunks further. Are you shipping ICU already separately (or at least the relevant data file(s))?  It should be possible to tell Flutter to link against that instead of embedding another copy.

**Comment by wangying3426** [comment]
> Sorry - I've been on leave and a bit slow to get back to things.
> 
> Using your own libskia may or may not introduce issues - Flutter uses at least one instance of non-public API from Skia (for base32 decoding ?!), and may use others. Even for the public API, Flutter tends to build against the very latest revision of Skia, which sometimes has breaking changes to API that get fixed in Flutter (for example, when a Skia method changes its signature).
> 
> It's definitely possible to build Skia separately as a dynamic library, but it would likely be important to use the _exact same revision as what Flutter expects_ in the rest of your application, which may or may not help you here.
> 
> It would be interesting to have a way to let dart:io delegate to some other implementation rather than the Dart VM's. I made comments more about this in #32161. I think to make it really work, you'd want to covery the entire API surface of dart:io w.r.t. Sockets and HTTP calls though - so that any other plugins or framework bits you use that rely on it will still work.
> 
> ICU has already been heavily optimized for Flutter - we've cut out a lot that we didn't need (/cc @goderbauer). There are additional plans to see if we can compress some of the larger chunks further. Are you shipping ICU already separately (or at least the relevant data file(s))? It should be possible to tell Flutter to link against that instead of embedding another copy.

We have no progress on the icu library now.  ICU, Minikin and HarfBuzz are tightly coupled to each other, so it seems difficult to remove any of them separately.

In fact, we have other rendering engine which also used harfbuzz and icu  in our apps, therefore it will be helpful to reduce the package size If they can be reused.

**Comment by dnfield** [comment]
Filed https://github.com/flutter/flutter/issues/43513 to explore the ICU/harfbuzz area.

**Comment by xster** [comment]
Interesting write-up from ByteDance https://translate.google.com/translate?hl=&sl=zh-CN&tl=en&u=https%3A%2F%2Fmp.weixin.qq.com%2Fs%2FLs3cDcqjlyOX80PXUO0wRw&sandbox=1

**Comment by xster** [comment]
Also worth cross-referencing @wangying3426's analysis in https://github.com/dart-lang/sdk/issues/38647 for potential customizability.

**Comment by xster** [comment]
A relevant extraction from the ByteDance article (via google translate).

> The first part, removing boringSSL, can use the method channel to call the source network library to replace the Dart Http function. Just as we basically never use OKHttp naked on Android, we have to do some dynamic routing and reconnection. There are also various targeted optimizations for the domestic network. Dart's native network library performance must be better than the native optimized network library for the domestic environment. At this time, we can use the Method Channel to call the source network library. It replaces the Dart Http function, so the performance will definitely be improved, it will not be reduced, but it will also bring the benefits of package volume.

Would be good to do this while providing synergy with the feature request to [support cronet as the networking implementation](https://github.com/flutter/flutter/issues/17413)

**Comment by wangying3426** [comment]
Hello，@xster @dnfield , first of all I'd like to report what we have done：
-  Removed BoringSSL，and will check if  "SecureSocket" class is misused at compile time on  release mode.
-  Compiled Skia as a dynamic library, and shared it with another game engine, and controlled "export" to reduce symbol export. Also, removed parts unused features of Skia, some pathops etc.
-  Removed parts of  ICU international data resources, burmesedict.dict,thaidict.dict etc.
-  Removed JPEG and PNG decoders on iOS,  and used Core Graphics framework to fallback.

Because of our flutter applications will shared some dynamic libraries with another game engine, what we hope to modularize the three-party libraries as follows:
-  Skia and BoringSSL, which are most important, needed to be shared. P0 priority.
-  It will be very helpful if the ICU, HarfBuzz and FreeType2 can be shared too. P1 priority.
-  WEBP, PNG and JPEG decoders also need to be shared, but is not too important because they take up a small amount of package size. P2 priority.

**Comment by mimbrown** [comment]
> What are your primary goals? For example, are you most concerned with reducing disk footprint of the engine, reducing latency of startup at runtime, including features not currently bundled into the engine, or something else?

@dnfield I have a feature I'm looking to include. I'd like to enable Graphite2 in Harfbuzz (the `--with-graphite2` flag while configuring). How doable is this at the moment?

**Comment by Hixie** [comment]
I'm going to close this issue because it describes a solution rather than a problem. I would encourage those of you interested in this issue to file new bugs that describe the reason why you were interested in this issue (e.g. is it so that you can add more modules? so that you can sell plugins? so that you can remove modules to get a smaller engine size? if so, how small is small enough? etc).
